src/main/java/com/example/online_shop_api/Service/Products/ProductService.java

```diff
@@ -0,0 +1,79 @@
+import com.example.online_shop_api.Exceptions.ProductNotFoundException;
+import com.example.online_shop_api.Repository.Products.ProductRepository;
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import org.modelmapper.ModelMapper;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ProductService {
+
+  private final ProductRepository productRepository;
+  private final ModelMapper modelMapper;
+
+  public ResponseEntity<List<ProductResponseDto>> getAllProducts() {
+    List<Product> productList = productRepository.findAll();
+    return ResponseEntity.ok(
+        productList.stream()
+            .map(product -> modelMapper.map(product, ProductResponseDto.class))
+            .toList());
+  }
+
+  public ProductResponseDto getById(Long id) {
+    Product product =
+        productRepository.findById(id).orElseThrow(() -> new ProductNotFoundException(id));
+    return modelMapper.map(product, ProductResponseDto.class);
+  }
+
+  public ResponseEntity<ProductResponseDto> create(ProductCreationRequestDto request) {
+    Product product = mapToSpecificProduct(request);
+    productRepository.save(product);
+    return ResponseEntity.ok(modelMapper.map(product, ProductResponseDto.class));
+  }
+
+  public ResponseEntity<ProductResponseDto> update(ProductCreationRequestDto request, Long id) {
+    Product existingProduct =
+        productRepository.findById(id).orElseThrow(() -> new ProductNotFoundException(id));
+
+    Product updatedProduct = mapToSpecificProduct(request);
+    updatedProduct.setId(existingProduct.getId());
+
+    updatedProduct = productRepository.save(updatedProduct);
+
+    return ResponseEntity.ok(modelMapper.map(updatedProduct, ProductResponseDto.class));
+    }
+
+  public void delete(Long id) {
+    Product product =
+        productRepository.findById(id).orElseThrow(() -> new ProductNotFoundException(id));
+    productRepository.delete(product);
+  }
+  private Product mapToSpecificProduct(ProductCreationRequestDto request) {
+    String productType = normalizeProductType(request.getProductType());
+    request.setProductType(productType);
+
+    ProductRequestDto productRequestDto = request.getProductRequestDto();
+    return switch (productType) {
+      case "Accessory" -> modelMapper.map(productRequestDto, Accessory.class);
+      case "Decoration" -> modelMapper.map(productRequestDto, Decoration.class);
+      case "Drink" -> modelMapper.map(productRequestDto, Drink.class);
+      case "Food" -> modelMapper.map(productRequestDto, Food.class);
+      case "Others" -> modelMapper.map(productRequestDto, Others.class);
+      case "Railing" -> modelMapper.map(productRequestDto, Railing.class);
+      case "Sanitary" -> modelMapper.map(productRequestDto, Sanitary.class);
+      default -> throw new IllegalArgumentException("Invalid product type: " + productType);
+    };
```

Comment:
Does this mapper add all the fields for the mapped product type? For example - we give it a productType of Railing, does the Mapper include and return an object having the fields material, color and brand? If yes - perfect.

Reply:
yes this method returns the object with the specific fields it has